synapse/handlers/federation.py

```diff
@@ -281,7 +281,7 @@ async def on_receive_pdu(self, origin, pdu, sent_to_us_directly=False) -> None:
                             raise Exception(
                                 "Error fetching missing prev_events for %s: %s"
                                 % (event_id, e)
-                            )
+                            ) from e
```

Comment:
the idea here is to change this:

```
2020-09-21 06:01:39,826 - synapse.federation.federation_server - 309 - ERROR - PUT-1098178-$s5rml0MNFzN0HCySGnCzRn1CXnGiIvoQCfYtL68-ucI- Failed to handle PDU $s5rml0MNFzN0HCySGnCzRn1CXnGiIvoQCfYtL68-ucI
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 276, in on_receive_pdu
    origin, pdu, prevs, min_depth
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 540, in _get_missing_events_for_pdu
    await self.on_receive_pdu(origin, ev, sent_to_us_directly=False)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 418, in on_receive_pdu
    await self._process_received_pdu(origin, pdu, state=state)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 701, in _process_received_pdu
    context = await self._handle_new_event(origin, event, state=state)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 1901, in _handle_new_event
    [(event, context)], backfilled=backfilled
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 2918, in persist_events_and_notify
    backfilled=backfilled,
  File "/usr/local/lib/python3.7/site-packages/synapse/replication/http/_base.py", line 196, in send_request
    result = await request_func(uri, data, headers=headers)
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 430, in post_json_get_json
    "POST", uri, headers=Headers(actual_headers), data=json_str
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 332, in request
    response = await make_deferred_yieldable(request_deferred)
twisted.web._newclient.ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/synapse/federation/federation_server.py", line 298, in process_pdus_for_room
    await self._handle_received_pdu(origin, pdu)
  File "/usr/local/lib/python3.7/site-packages/synapse/federation/federation_server.py", line 683, in _handle_received_pdu
    await self.handler.on_receive_pdu(origin, pdu, sent_to_us_directly=True)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 281, in on_receive_pdu
    % (event_id, e)
Exception: Error fetching missing prev_events for $s5rml0MNFzN0HCySGnCzRn1CXnGiIvoQCfYtL68-ucI: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]
```

Into this:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 276, in on_receive_pdu
    origin, pdu, prevs, min_depth
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 540, in _get_missing_events_for_pdu
    await self.on_receive_pdu(origin, ev, sent_to_us_directly=False)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 418, in on_receive_pdu
    await self._process_received_pdu(origin, pdu, state=state)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 701, in _process_received_pdu
    context = await self._handle_new_event(origin, event, state=state)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 1901, in _handle_new_event
    [(event, context)], backfilled=backfilled
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 2918, in persist_events_and_notify
    backfilled=backfilled,
  File "/usr/local/lib/python3.7/site-packages/synapse/replication/http/_base.py", line 196, in send_request
    result = await request_func(uri, data, headers=headers)
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 430, in post_json_get_json
    "POST", uri, headers=Headers(actual_headers), data=json_str
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 332, in request
    response = await make_deferred_yieldable(request_deferred)
twisted.web._newclient.ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/synapse/federation/federation_server.py", line 298, in process_pdus_for_room
    await self._handle_received_pdu(origin, pdu)
  File "/usr/local/lib/python3.7/site-packages/synapse/federation/federation_server.py", line 683, in _handle_received_pdu
    await self.handler.on_receive_pdu(origin, pdu, sent_to_us_directly=True)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/federation.py", line 281, in on_receive_pdu
    % (event_id, e)
Exception: Error fetching missing prev_events for $s5rml0MNFzN0HCySGnCzRn1CXnGiIvoQCfYtL68-ucI: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]
```

it's maybe marginally clearer?